File: Algorithm/BinarySearch/leetcode/33. Search in Rotated Sorted Array/Search-in-Rotated-Sorted-Array-claude.ts
// この問題は「Search in Rotated Sorted Array」という典型的な二分探索の応用問題ですね。O(log n)の時間計算量が要求されているので、二分探索を使用します。

// 回転されたソート済み配列の特徴を利用して解決します：
// - 配列を半分に分けると、必ずどちらか一方は完全にソートされている
// - ソートされている側でtargetが範囲内にあるかチェックし、探索範囲を絞り込むこのソリューションのポイント：

// **アルゴリズムの解説：**
// 1. 標準的な二分探索をベースに、回転された配列の特性を活用
// 2. 配列を中央で分割すると、必ずどちらか一方は完全にソートされている
// 3. ソートされている側でtargetが範囲内にあるかを判定し、探索範囲を絞り込む

// **時間・空間効率：**
// - **時間計算量**: O(log n) - 二分探索により毎回探索範囲を半分に削減
// - **空間計算量**: O(1) - 追加の配列やデータ構造を使用せず、定数個の変数のみ

// **最適化のポイント：**
// 1. `Math.floor((left + right) / 2)`でオーバーフローを回避
// 2. 条件分岐を最小限に抑えて処理効率を向上
// 3. 不要な変数宣言を避けてメモリ使用量を削減

// この実装はLeetCodeの制約（配列長5000以下、値の範囲-10^4から10^4）に対して十分効率的で、実行時間・メモリ使用量ともに最適化されています。

/**
 * 回転されたソート済み配列から指定された値を検索する
 * @param nums - 回転された可能性のある昇順ソート済み配列（重複なし）
 * @param target - 検索対象の整数値
 * @returns targetが存在する場合はそのインデックス、存在しない場合は-1
 * 
 * 時間計算量: O(log n) - 二分探索を使用
 * 空間計算量: O(1) - 定数の追加メモリのみ使用
 */
function search(nums: number[], target: number): number {
    let left = 0;
    let right = nums.length - 1;
    
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        
        // targetが見つかった場合
        if (nums[mid] === target) {
            return mid;
        }
        
        // 左半分がソートされている場合
        if (nums[left] <= nums[mid]) {
            // targetが左半分の範囲内にある場合
            if (nums[left] <= target && target < nums[mid]) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        } 
        // 右半分がソートされている場合
        else {
            // targetが右半分の範囲内にある場合
            if (nums[mid] < target && target <= nums[right]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
    }
    
    return -1;
}

// テストケース
console.log(search([4,5,6,7,0,1,2], 0)); // Expected: 4
console.log(search([4,5,6,7,0,1,2], 3)); // Expected: -1
console.log(search([1], 0)); // Expected: -1
console.log(search([1], 1)); // Expected: 0
console.log(search([1,3], 3)); // Expected: 1

File: Algorithm/BinarySearch/leetcode/33. Search in Rotated Sorted Array/Search-in-Rotated-Sorted-Array-claude.py
# LeetCodeのPython形式で、型安全かつ効率的な実装を提供します。この実装の特徴とPythonでの最適化ポイントを説明します：

# ## 🐍 **Python特有の最適化**

# ### 1. **型安全性の確保**
# ```python
# from typing import List

# def search(self, nums: List[int], target: int) -> int:
# ```
# - `typing.List`を使用してPylanceの型エラーを回避
# - 全ての変数に明示的な型注釈を追加
# - 関数の引数と戻り値の型を明確に定義

# ### 2. **メモリ効率の最適化**
# ```python
# left: int = 0
# right: int = len(nums) - 1
# mid: int = (left + right) // 2
# ```
# - 必要最小限の変数のみ使用（O(1)空間）
# - Pythonの整数演算は自動的にオーバーフロー対策済み
# - `//`演算子で整数除算を明示

# ### 3. **処理時間の最適化**
# - **条件分岐の最適化**: 最も頻繁なケースを先に判定
# - **短絡評価の活用**: `nums[left] <= target < nums[mid]`
# - **不要な計算の排除**: 毎回の`len(nums)`呼び出しを避ける

# ### 4. **LeetCode形式への準拠**
# ```python
# class Solution:
#     def search(self, nums: List[int], target: int) -> int:
# ```
# - 標準的なLeetCodeクラス構造
# - メソッド名は問題仕様に合わせて`search`
# - 引数順序も仕様通り

# ## 📊 **計算量の詳細分析**

# ### 時間計算量: **O(log n)**
# - 二分探索により毎回探索範囲を半分に削減
# - 最悪ケース: `⌊log₂(n)⌋ + 1`回の反復

# ### 空間計算量: **O(1)**
# - 追加配列不要
# - 変数3個のみ（`left`, `right`, `mid`）

# ## 🚀 **Python固有の利点**

# 1. **自動メモリ管理**: ガベージコレクションによる効率的なメモリ使用
# 2. **整数オーバーフロー耐性**: 任意精度整数による安全性
# 3. **簡潔な構文**: 可読性とパフォーマンスの両立
# 4. **型チェック**: 実行時エラーの事前検出

# この実装は、LeetCodeの制約下で最適なパフォーマンスを発揮し、Pylanceによる静的解析でもエラーが発生しない堅牢なコードです。

from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        """
        回転されたソート済み配列から指定された値を検索する
        
        Args:
            nums: 回転された可能性のある昇順ソート済み配列（重複なし）
            target: 検索対象の整数値
            
        Returns:
            targetが存在する場合はそのインデックス、存在しない場合は-1
            
        時間計算量: O(log n) - 二分探索を使用
        空間計算量: O(1) - 定数の追加メモリのみ使用
        """
        left: int = 0
        right: int = len(nums) - 1
        
        while left <= right:
            # オーバーフローを避けるための中央値計算
            mid: int = (left + right) // 2
            
            # targetが見つかった場合
            if nums[mid] == target:
                return mid
            
            # 左半分がソートされている場合
            if nums[left] <= nums[mid]:
                # targetが左半分の範囲内にある場合
                if nums[left] <= target < nums[mid]:
                    right = mid - 1
                else:
                    left = mid + 1
            # 右半分がソートされている場合
            else:
                # targetが右半分の範囲内にある場合
                if nums[mid] < target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        
        return -1


def test_solution() -> None:
    """
    テストケースの実行
    
    Returns:
        None
    """
    solution = Solution()
    
    # テストケース1: Example 1
    nums1: List[int] = [4, 5, 6, 7, 0, 1, 2]
    target1: int = 0
    result1: int = solution.search(nums1, target1)
    print(f"Test 1: nums={nums1}, target={target1}")
    print(f"Expected: 4, Got: {result1}")
    assert result1 == 4
    
    # テストケース2: Example 2
    nums2: List[int] = [4, 5, 6, 7, 0, 1, 2]
    target2: int = 3
    result2: int = solution.search(nums2, target2)
    print(f"Test 2: nums={nums2}, target={target2}")
    print(f"Expected: -1, Got: {result2}")
    assert result2 == -1
    
    # テストケース3: Example 3
    nums3: List[int] = [1]
    target3: int = 0
    result3: int = solution.search(nums3, target3)
    print(f"Test 3: nums={nums3}, target={target3}")
    print(f"Expected: -1, Got: {result3}")
    assert result3 == -1
    
    # 追加テストケース: 単一要素で一致する場合
    nums4: List[int] = [1]
    target4: int = 1
    result4: int = solution.search(nums4, target4)
    print(f"Test 4: nums={nums4}, target={target4}")
    print(f"Expected: 0, Got: {result4}")
    assert result4 == 0
    
    # 追加テストケース: 回転していない配列
    nums5: List[int] = [1, 2, 3, 4, 5]
    target5: int = 3
    result5: int = solution.search(nums5, target5)
    print(f"Test 5: nums={nums5}, target={target5}")
    print(f"Expected: 2, Got: {result5}")
    assert result5 == 2
    
    # 追加テストケース: 境界値（最小値を探索）
    nums6: List[int] = [3, 4, 5, 1, 2]
    target6: int = 1
    result6: int = solution.search(nums6, target6)
    print(f"Test 6: nums={nums6}, target={target6}")
    print(f"Expected: 3, Got: {result6}")
    assert result6 == 3
    
    print("\n✅ All tests passed!")


def analyze_complexity() -> None:
    """
    アルゴリズムの計算量解析
    
    Returns:
        None
    """
    print("📊 計算量解析:")
    print("⏱️  時間計算量: O(log n)")
    print("   - 二分探索により毎回探索範囲を半分に削減")
    print("   - 最悪ケースでも log₂(n) 回の反復で完了")
    print()
    print("💾 空間計算量: O(1)")
    print("   - left, right, mid の3つの変数のみ使用")
    print("   - 入力配列のサイズに依存しない定数空間")
    print()
    print("🚀 最適化ポイント:")
    print("   1. (left + right) // 2 でオーバーフロー対策")
    print("   2. 型注釈によるPylanceエラー回避")
    print("   3. 条件分岐の最適化で分岐予測の向上")
    print("   4. 不要な変数宣言を避けてメモリ効率化")


def explain_algorithm() -> None:
    """
    アルゴリズムの動作原理の説明
    
    Returns:
        None
    """
    print("🔍 アルゴリズムの核心:")
    print("回転されたソート済み配列の重要な性質:")
    print("→ 配列を任意の点で半分に分けると、必ずどちらか一方は完全にソートされている")
    print()
    print("例: [4,5,6,7,0,1,2] を mid=3 で分割")
    print("左半分: [4,5,6,7] ← ソート済み ✅")
    print("右半分: [0,1,2]   ← ソート済み ✅")
    print()
    print("探索戦略:")
    print("1. ソートされている側を特定 (nums[left] <= nums[mid])")
    print("2. その側でtargetが範囲内かチェック")
    print("3. 範囲内なら其の側を探索、範囲外なら反対側を探索")
    print("4. 探索範囲を半分に削減して繰り返し")


if __name__ == "__main__":
    explain_algorithm()
    print("\n" + "="*50 + "\n")
    analyze_complexity()
    print("\n" + "="*50 + "\n")
    test_solution()